kazoo/recipe/cache.py

```diff
@@ -0,0 +1,382 @@
+                    full_path = os.path.join(path, child)
+                    if child not in self._children:
+                        node = TreeNode(self._tree, full_path, self)
+                        self._children[child] = node
+                        node.was_created()
+        elif method_name == 'get':
+            try:
+                data, stat = result.get()
+            except NoNodeError:
+                self.was_deleted()
+            else:
+                old_data, self._data = (
+                    self._data, NodeData.make(path, data, stat))
+
+                old_state, self._state = self._state, self.STATE_LIVE
+                if old_state == self.STATE_LIVE:
+                    if old_data is None or old_data.stat.mzxid != stat.mzxid:
+                        self._publish_event(TreeEvent.NODE_UPDATED, self._data)
+                else:
+                    self._publish_event(TreeEvent.NODE_ADDED, self._data)
+        else:  # pragma: no cover
+            logger.warning('unknown operation %s' % method_name)
+            self._tree._outstanding_ops -= 1
+            return
+
+        self._tree._outstanding_ops -= 1
+        if self._tree._outstanding_ops == 0 and not self._tree._is_initialized:
+            self._tree._is_initialized = True
+            self._publish_event(TreeEvent.INITIALIZED)
+
+
+class TreeEvent(tuple):
+    """The immutable event tuple of cache."""
+
+    NODE_ADDED = 0
+    NODE_UPDATED = 1
+    NODE_REMOVED = 2
+    CONNECTION_SUSPENDED = 3
+    CONNECTION_RECONNECTED = 4
+    CONNECTION_LOST = 5
+    INITIALIZED = 6
+
+    #: An enumerate integer to indicate event type.
+    event_type = property(operator.itemgetter(0))
+
+    #: A :class:`~kazoo.recipe.cache.NodeData` instance.
+    event_data = property(operator.itemgetter(1))
+
+    @classmethod
+    def make(cls, event_type, event_data):
+        """Creates a new TreeEvent tuple.
+
+        :returns: A :class:`~kazoo.recipe.cache.TreeEvent` instance.
+        """
+        assert event_type in (
+            cls.NODE_ADDED, cls.NODE_UPDATED, cls.NODE_REMOVED,
+            cls.CONNECTION_SUSPENDED, cls.CONNECTION_RECONNECTED,
+            cls.CONNECTION_LOST, cls.INITIALIZED)
+        return cls((event_type, event_data))
+
+
+class NodeData(tuple):
+    """The immutable node data tuple of cache."""
+
+    #: The absolute path string of current node.
+    path = property(operator.itemgetter(0))
+
+    #: The bytes data of current node.
+    data = property(operator.itemgetter(1))
+
+    #: The stat information of current node.
+    stat = property(operator.itemgetter(2))
+
+    @classmethod
+    def make(cls, path, data, stat):
+        """Creates a new NodeData tuple.
+
+        :returns: A :class:`~kazoo.recipe.cache.NodeData` instance.
+        """
+        return cls((path, data, stat))
+
+
+@contextlib.contextmanager
+def handle_exception(listeners):
+    try:
+        yield
+    except Exception as e:
```

Comment:
If we ask people don't throw any exception except `KazooException` in their listeners, people must remember to write `try: ... except Exception: ...` in all listeners, or they will have unexpected behavior. It looks weird.

I don't like the "catch and hide" style too, especially for unknown exceptions. But we have `logger.exception` here. I believe it will leave enough information for unexpected exceptions from listeners.